pkg/pgxconn/implement.go

```diff
@@ -0,0 +1,105 @@
+package pgxconn
+
+import (
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgproto3/v2"
+	"github.com/jackc/pgx/v4"
+)
+
+// rowWithMetric wraps the row returned by QueryRow() for metrics telemetry.
+type rowWithMetric struct {
+	r pgx.Row
+}
+
+func (w rowWithMetric) Scan(dest ...interface{}) error {
+	err := w.r.Scan(dest...)
+	if err != nil && err != pgx.ErrNoRows {
+		errorsTotal.With(promMethodLabel("query_row")).Inc()
+	}
+	return err
+}
+
+// rowsWithDuration wraps the rows and records the duration of the batch till it closed.
+type rowsWithDuration struct {
+	r     pgx.Rows
+	start time.Time
+}
+
+func newRowsWithDuration(r pgx.Rows) rowsWithDuration {
+	return rowsWithDuration{
+		r:     r,
+		start: time.Now(),
+	}
+}
+
+func (w rowsWithDuration) Close() {
+	duration.With(promMethodLabel("query")).Observe(time.Since(w.start).Seconds())
+	w.r.Close()
+}
+
+func (w rowsWithDuration) Err() error {
+	return w.r.Err()
+}
+
+func (w rowsWithDuration) CommandTag() pgconn.CommandTag {
+	return w.r.CommandTag()
+}
+
+func (w rowsWithDuration) FieldDescriptions() []pgproto3.FieldDescription {
+	return w.r.FieldDescriptions()
+}
+
+func (w rowsWithDuration) Next() bool {
+	return w.r.Next()
+}
+
+func (w rowsWithDuration) Scan(dest ...interface{}) error {
+	return w.r.Scan(dest...)
+}
+
+func (w rowsWithDuration) Values() ([]interface{}, error) {
+	return w.r.Values()
+}
+
+func (w rowsWithDuration) RawValues() [][]byte {
+	return w.r.RawValues()
+}
+
+// batchResultsWithDuration wraps the SendBatch and records the duration of the batch till it closed.
+type batchResultsWithDuration struct {
+	b     pgx.BatchResults
+	start time.Time
+}
+
+func newBatchResultsWithDuration(b pgx.BatchResults) batchResultsWithDuration {
+	return batchResultsWithDuration{
+		b:     b,
+		start: time.Now(),
+	}
+}
+
+func (w batchResultsWithDuration) Close() error {
+	defer func() {
+		duration.With(promMethodLabel("send_batch"))
+	}()
+	return w.b.Close()
+}
+
+func (w batchResultsWithDuration) Exec() (pgconn.CommandTag, error) {
+	return w.b.Exec()
+}
+
+func (w batchResultsWithDuration) Query() (pgx.Rows, error) {
+	rows, err := w.b.Query()
+	return newRowsWithDuration(rows), err
```

Comment:
This does not seem correct as the `Query` is executed above thus `newRowsWithDuration` will not include that time.

Reply:
You might want to do something like `newRowsWithDuration(w.b.Query())...`

Reply:
No, because Query() results pgxRows, which when iterated gives the results in lazy loading manner.

Reply:
Hmm I would not say so by looking into implementation here https://github.com/jackc/pgx/blob/master/conn.go#L547 . It seems that it will prepare a statement and execute it